File: create_validate_tests.py
"""
Helper script to write `validate` methods for transactions.
"""
import sys
import json


NORMAL_TYPES = ["number", "string"]
NUMBERS = ["0", "1"]

BINARY_CODEC_FIXTURES = "./packages/ripple-binary-codec/test/fixtures/codec-fixtures.json"

def get_tx(tx_name):
    with open(BINARY_CODEC_FIXTURES) as f:
        fixtures = json.load(f)
        transactions = fixtures["transactions"]
        print(len(transactions))
        valid_txs = [tx["json"] for tx in transactions if tx["json"]["TransactionType"] == tx_name]
        valid_tx = valid_txs[0]
        del valid_tx["TxnSignature"]
        del valid_tx["SigningPubKey"]
        return json.dumps(valid_tx, indent=2)


def main():
    model_name = sys.argv[1]
    filename = f"./packages/xrpl/src/models/transactions/{model_name}.ts"
    model, tx_name = get_model(filename)
    return process_model(model, tx_name)


# Extract just the model from the file
def get_model(filename):
    model = ""
    started = False
    ended = False
    with open(filename) as f:
        for line in f:
            if ended:
                continue
            if not started and not line.startswith("export"):
                continue
            if not started and "Flags" in line:
                continue
            if not started:
                started = True
            model += line
            if line == '}\n':
                ended = True

    lines = model.split("\n")
    name_line = lines[0].split(" ")
    tx_name = name_line[2]
    return model, tx_name


def get_invalid_value(param_types):
    if len(param_types) == 1:
        param_type = param_types[0]
        if param_type == "number":
            return "'number'"
        elif param_type == "string":
            return 123
        elif param_type == "IssuedCurrency":
            return {"test": "test"}
        elif param_type == "Amount":
            return {"currency": "ETH"}
        elif param_type == "XChainBridge":
            return {"XChainDoor": "test"}
        else:
            raise Exception(f"{param_type} not supported yet")

    simplified_param_types = {param_types[i] for i in range(len(param_types)) if i % 2 == 0}
    if simplified_param_types == {'0', '1'}:
        return 2
    elif simplified_param_types == {"number", "string"}:
        return {"currency": "ETH"}
    else:
        raise Exception(f"{simplified_param_types} not supported yet")

# Process the model and build the tests

def process_model(model, tx_name):
    output = ""

    for line in model.split("\n"):
        # skip the things we don't want to test
        if line == "":
            continue
        if line.startswith("export"):
            continue
        if line == "}":
            continue
        line = line.strip()

        if line.startswith("TransactionType"):
            continue
        if line.startswith("Flags"):
            # TODO: support flag checking
            continue

        split = line.split(" ")
        param = split[0].strip("?:")
        param_types = split[1:]
        optional = split[0].endswith("?:")

        if not optional:
            output += f"""  it(`throws w/ missing {param}`, function () {{
    delete tx.{param}

    assert.throws(
      () => validate{tx_name}(tx),
      ValidationError,
      '{tx_name}: missing field {param}',
    )
    assert.throws(
      () => validate(tx),
      ValidationError,
      '{tx_name}: missing field {param}',
    )
  }})

"""

        fake_value = get_invalid_value(param_types)
        output += f"""  it(`throws w/ invalid {param}`, function () {{
    tx.{param} = {fake_value}

    assert.throws(
      () => validate{tx_name}(tx),
      ValidationError,
      '{tx_name}: invalid field {param}',
    )
    assert.throws(
      () => validate(tx),
      ValidationError,
      '{tx_name}: invalid field {param}',
    )
  }})

"""

    output = output[:-2]
    output += "})\n"

    output = f"""import {{ assert }} from 'chai'

import {{ validate, ValidationError }} from '../../src'
import {{ validate{tx_name} }} from '../../src/models/transactions/{tx_name}'

/**
 * {tx_name} Transaction Verification Testing.
 *
 * Providing runtime verification testing for each specific transaction type.
 */
describe('{tx_name}', function () {{
  let tx

  beforeEach(function () {{
    tx = {get_tx(tx_name)} as any
  }})

  it('verifies valid {tx_name}', function () {{
    assert.doesNotThrow(() => validate{tx_name}(tx))
    assert.doesNotThrow(() => validate(tx))
  }})

""" + output

    return output


if __name__ == "__main__":
    print(main())

File: packages/xrpl/tools/createValidateTests.js
const fs = require('fs')
const path = require('path')
const fixtures = require('../../ripple-binary-codec/test/fixtures/codec-fixtures.json')

const NORMAL_TYPES = ['number', 'string']
const NUMBERS = ['0', '1']

function getTx(txName) {
  transactions = fixtures.transactions
  const validTxs = fixtures.transactions
    .filter((tx) => tx.json.TransactionType === txName)
    .map((tx) => tx.json)
  if (validTxs.length == 0) {
    throw new Error(`Must have ripple-binary-codec fixture for ${txName}`)
  }
  const validTx = validTxs[0]
  delete validTx.TxnSignature
  delete validTx.SigningPubKey
  return JSON.stringify(validTx, null, 2)
}

function main(modelName) {
  const filename = path.join(
    path.dirname(__filename),
    `../src/models/transactions/${modelName}.ts`,
  )
  const [model, txName] = getModel(filename)
  return processModel(model, txName)
}

// Extract just the model from the file
function getModel(filename) {
  let model = ''
  let started = false
  let ended = false
  const data = fs.readFileSync(filename, 'utf8')
  const lines = data.split('\n')
  for (let line of lines) {
    if (ended) {
      continue
    }
    if (!started && !line.startsWith('export')) {
      continue
    }
    if (!started && line.includes('Flags')) {
      continue
    }
    if (!started) {
      started = true
    }
    model += line + '\n'
    if (line === '}') {
      ended = true
    }
  }
  const name_line = model.split('\n')[0].split(' ')
  const txName = name_line[2]
  return [model, txName]
}

function getInvalidValue(paramTypes) {
  if (paramTypes.length === 1) {
    const paramType = paramTypes[0]
    if (paramType == 'number') {
      return "'number'"
    } else if (paramType == 'string') {
      return 123
    } else if (paramType == 'IssuedCurrency') {
      return JSON.stringify({ test: 'test' })
    } else if (paramType == 'Currency') {
      return JSON.stringify({ test: 'test' })
    } else if (paramType == 'Amount') {
      return JSON.stringify({ currency: 'ETH' })
    } else if (paramType == 'XChainBridge') {
      return JSON.stringify({ XChainDoor: 'test' })
    } else {
      throw Error(`${paramType} not supported yet`)
    }
  }

  const simplifiedParamTypes = paramTypes.filter(
    (_paramType, index) => index % 2 == 0,
  )
  if (JSON.stringify(simplifiedParamTypes) === '["0","1"]') {
    return 2
  } else if (JSON.stringify(simplifiedParamTypes) === '["number","string"]') {
    return JSON.stringify({ currency: 'ETH' })
  } else {
    throw Error(`${simplifiedParamTypes} not supported yet`)
  }
}

// Process the model and build the tests

function processModel(model, txName) {
  let output = ''
  for (let line of model.split('\n')) {
    if (line == '') {
      continue
    }
    if (line.startsWith('export')) {
      continue
    }
    if (line == '}') {
      continue
    }
    line = line.trim()
    if (line.startsWith('TransactionType')) {
      continue
    }
    if (line.startsWith('Flags')) {
      // TODO: support flag checking
      continue
    }
    if (line.startsWith('/**')) {
      continue
    }
    if (line.startsWith('*')) {
      continue
    }

    const split = line.split(' ')
    const param = split[0].replace('?:', '').replace(':', '').trim()
    const paramTypes = split.slice(1)
    const optional = split[0].endsWith('?:')

    if (!optional) {
      output += `  it("throws w/ missing ${param}", function () {
    delete tx.${param}

    assert.throws(
      () => validate${txName}(tx),
      ValidationError,
      '${txName}: missing field ${param}',
    )
    assert.throws(
      () => validate(tx),
      ValidationError,
      '${txName}: missing field ${param}',
    )
  })

`
    }

    const fakeValue = getInvalidValue(paramTypes)
    output += `  it('throws w/ invalid ${param}', function () {
    tx.${param} = ${fakeValue}

    assert.throws(
      () => validate${txName}(tx),
      ValidationError,
      '${txName}: invalid field ${param}',
    )
    assert.throws(
      () => validate(tx),
      ValidationError,
      '${txName}: invalid field ${param}',
    )
  })

`
  }

  output = output.substring(0, output.length - 2)
  output += '\n})\n'
  output =
    `import { assert } from 'chai'

import { validate, ValidationError } from '../../src'
import { validate${txName} } from '../../src/models/transactions/${txName}'

/**
 * ${txName} Transaction Verification Testing.
 *
 * Providing runtime verification testing for each specific transaction type.
 */
describe('${txName}', function () {
  let tx

  beforeEach(function () {
    tx = ${getTx(txName)} as any
  })

  it('verifies valid ${txName}', function () {
    assert.doesNotThrow(() => validate${txName}(tx))
    assert.doesNotThrow(() => validate(tx))
  })

` + output

  return output
}

if (require.main === module) {
  if (process.argv.length < 3) {
    console.log(`Usage: ${process.argv[0]} ${process.argv[1]} TxName`)
    process.exit(1)
  }
  console.log(main(process.argv[2]))
}

module.exports = main
